README.md
| :exclamation: The v1 API has been deprecated, endpoints are no longer available! Use v2 instead! |
| ------------------------------------------------------------------------------------------------ |


# Minecraft Server Properties API

Find useful information about Minecraft's Java Edition server configuration file, server.properties, with this simple, RESTful API!

## Usage

Make a simple GET request and decode the JSON response!

To get the whole documentation: `GET http://api.mcbonanza.games//v1serverproperties`  
To get a single key: `GET http://api.mcbonanza.games/serverproperties/v1/{key}`, where `{key}` is a valid server.properties key.  
To get metadata (such as the default limit value and the possible value types): `GET http://api.mcbonanza.games/v1/serverproperties/meta/`

### Response examples

- **Whole documentation** (on `GET http://api.mcbonanza.games/v1/serverproperties`)

```json
{
  "options": {
    "contains": [],
    "type": [],
    "upcoming": ""
  },
  "properties": [
    {
      "name": "allow-flight",
      "type": "boolean",
      "defaultValue": "false",
      "values": {
        "min": 0,
        "max": 1,
        "possible": ["false", "true"]
      },
      "description": "Allows users to ...",
      "upcoming": false,
      "upcomingVersion": ""
    }
    ...
  ]
}
```

You can filter your requests with the provided options!

- `contains` option only returns the properties that contain the substring the key is attributed
- `type` option only returns the properties of the requested type
- `upcoming` option only returns properties that are going to be implemented in future versions, if `"true"`, or only properties currently available, if `"false"`

_Example request_: `GET http://api.mcbonanza.games/v1/serverproperties?upcoming=true&type=integer`

```json
{
  "options": {
    "contains": "",
    "type": "integer",
    "upcoming": "true"
  },
  "properties": [
    {
      "name": "entity-broadcast-range-percentage",
      "type": "integer",
      "defaultValue": "100",
      "values": {
        "min": 0,
        "max": 500,
        "possible": []
      },
      "description": "Controls how close ...",
      "upcoming": true,
      "upcomingVersion": "JE 1.16"
    }
  ]
}
```

If there is no property that matches your options, an empty array will be returned.

- **Single key** (on `GET http://api.mcbonanza.games/v1/serverproperties/difficulty`)

```json
{
  "name": "difficulty",
  "type": "string",
  "defaultValue": "easy",
  "values": {
    "min": -2147483648,
    "max": -2147483648,
    "possible": ["peaceful", "easy", "normal", "hard"]
  },
  "description": "Defines the difficulty ...",
  "upcoming": false,
  "upcomingVersion": ""
}
```

- **Error** (on `GET http://api.mcbonanza.games/v1/serverproperties/sarmale`). The errors are appended to the array in the order they occur.

```json
{
  "errors": [
    {
      "error": "404 Not Found, key \"sarmale\" doesn't exist",
      "retry": false
    }
  ]
}
```

## How it works

The API scrapes the official Minecraft Wiki page (specifically the [server.properties page](https://minecraft.gamepedia.com/Server.properties)), also evaluating
the mentioned values and limits a property can be assigned. The math expressions are evaluated using [math.js web service](https://api.mathjs.org/). Then the data is sent to the user, caching it if possible.

## Contributions

Fork this project, create a branch from develop, do your work, and open a pull request in the origin's develop branch.
Don't forget to get the module's dependencies! Use goimports to format the code and LF line endings.


CHANGELOG.md
# Changelog

## [1.2.2](https://github.com/tmaxmax/serverpropertiesapi/compare/1.2.1...1.2.2)

### Changed

- API endpoints make more sense
- API is served on https

### Fixed

- Unindented marshaled JSON from single property API response is now indented
- Accepted content-type checking now works for multiple accepted types

## [1.2.1](https://github.com/tmaxmax/serverpropertiesapi/compare/1.2.0...1.2.1)

### Added

- Filter values can now be comma-separated
- The responses are sent gzipped if the client supports it

### Changed

- Marshaled JSON is now indented

### Fixed

- Wrong header key for Content-Type fixed, was Content-Types
- Filter values are now validated, the API returning a 400 Bad Request error when they are invalid.

## [1.2.0](https://github.com/tmaxmax/serverpropertiesapi/compare/1.1.0...1.2.0)

### Added

- API endpoint for requesting from the whole documentation now supports options to filter the results (see [README.md](README.md))

### Changed

- API now scrapes the website on each request, if not cached (this ensures the information is always up to date)
- Requests are cached for 24 hours

## [1.1.0](https://github.com/tmaxmax/serverpropertiesapi/compare/1.0.1...1.1.0)

### Added

- API endpoint for requesting metadata, such as property types and limit default value
- Unsupported HTTP methods now have their own handler function

## [1.0.1](https://github.com/tmaxmax/serverpropertiesapi/compare/1.0.0...1.0.1)

### Added

- The project now has a changelog

### Changed

- Scraper ignores possible values for boolean properties, as they are redundant

### Fixed

- Scraper now picks up the possible values correctly

## [1.0.0](https://github.com/tmaxmax/serverpropertiesapi/releases/tag/1.0.0)

This is the first API release

### Added

- Implemented the Minecraft Gamepedia Wiki server.properties page scraper
- API endpoint for requesting from whole documentation
- API endpoint for requesting the documentation of a single key
- API error handling for requesting unsupported file formats and inexistent resources


api_test.go
package serverpropertiesapi

import (
	"encoding/json"
	"testing"
)

func TestGetServerProperties(t *testing.T) {
	prop, err := GetServerProperties()
	if err != nil {
		t.Errorf("Failed. Error: %v\n", err)
	}

	for i, p := range prop {
		if p.Name == "" || p.Type == "" || p.Default == "" || p.Description == "" {
			t.Errorf("\tFailed on property %d, one or more fields are empty. Check if website format hasn't changed!\nProperty instance: %+v\n", i, p)
		}
		if p.Values.Min == PropertyDefaultLimitValue && p.Values.Max != p.Values.Min {
			t.Errorf("\tFailed on property %d, invalid Min and Max.\np.Values.Min = %d, p.Values.Max = %d\n", i, p.Values.Min, p.Values.Max)
		}
		if (p.Upcoming && p.UpcomingVersion == "") || (!p.Upcoming && p.UpcomingVersion != "") {
			t.Errorf("\tFailed on property %d, Upcoming information mismatch.\np.Upcoming = %t, p.UpcomingVersion = %s", i, p.Upcoming, p.UpcomingVersion)
		}
	}

	jsonData, err := json.MarshalIndent(prop, "", "  ")
	if err != nil {
		t.Errorf("Failed, couldn't marshal JSON. Error: %v\n", err)
	}
	t.Logf("Succeeded. Marshaled JSON:\n%s\n", string(jsonData))
}


scraper_test.go
package serverpropertiesapi

import (
	"encoding/json"
	"testing"
)

func TestServerProperties(t *testing.T) {
	prop, err := ServerProperties(Options{})
	if err != nil {
		t.Errorf("Failed. Error: %v\n", err)
	}

	for i, p := range prop {
		// If any of these fields are empty, it means that gathering the information from the website failed.
		// This could be caused by a redesign, for example.
		if p.Name == "" || p.Type == "" || p.Default == "" || p.Description == "" {
			t.Errorf("\tFailed on property %d, one or more fields are empty. Check if website format hasn't changed!\nProperty instance: %+v\n", i, p)
		}
		// This is to assure that the limits have the default value if they are not mentioned in the documentation.
		if p.Values.Min == propertyDefaultLimitValue && p.Values.Max != p.Values.Min {
			t.Errorf("\tFailed on property %d, invalid Min and Max.\np.Values.Min = %d, p.Values.Max = %d\n", i, p.Values.Min, p.Values.Max)
		}
		// If the feature isn't going to be added in an upcoming version, but the UpcomingVersion string isn't empty,
		// or vice-versa, it obviously means that gathering the information failed.
		if (p.Upcoming && p.UpcomingVersion == "") || (!p.Upcoming && p.UpcomingVersion != "") {
			t.Errorf("\tFailed on property %d, Upcoming information mismatch.\np.Upcoming = %t, p.UpcomingVersion = %s", i, p.Upcoming, p.UpcomingVersion)
		}
	}

	jsonData, err := json.MarshalIndent(prop, "", "  ")
	if err != nil {
		t.Errorf("Failed, couldn't marshal JSON. Error: %v\n", err)
	}
	t.Logf("Succeeded. Marshaled JSON:\n%s\n", string(jsonData))
}


v2/scraper_test.go
package v2

import (
	"encoding/json"
	"testing"
)

func TestServerProperties(t *testing.T) {
	filters := []filters{
		{
			contains:  nil,
			exactName: "difficulty",
			sortMap:   nil,
			typesMap:  nil,
			upcoming:  "",
		},
		{
			contains:  nil,
			exactName: "",
			sortMap:   nil,
			typesMap:  nil,
			upcoming:  "",
		},
		{
			contains: map[string]bool{
				"allow": true,
			},
			exactName: "",
			sortMap:   nil,
			typesMap:  nil,
			upcoming:  "",
		},
		{
			contains: map[string]bool{
				"max":   false,
				"allow": false,
			},
			exactName: "",
			sortMap: map[string]bool{
				"name": true,
			},
			typesMap: nil,
			upcoming: "",
		},
		{
			contains:  nil,
			exactName: "",
			sortMap: map[string]bool{
				"name": false,
			},
			typesMap: map[string]bool{
				minecraftStringTypename: true,
			},
		},
	}
	for i, f := range filters {
		t.Logf("\nTest %d...\n", i)
		validateTypes(&f)
		prop, err := ServerProperties(&f)
		if err != nil {
			t.Errorf("\nTest %d failed, error: %v\n", i, err)
		}
		if f.exactName != "" && (len(prop) > 1 || prop[0].Name != f.exactName) {
			t.Errorf("exactName filtering failed")
		}
		for j, p := range prop {
			// If any of these fields are empty, it means that gathering the information from the website failed.
			// This could be caused by a redesign, for example.
			if p.Name == "" || p.Type == "" || p.Values.Default == "" || p.Description == "" {
				t.Errorf("\tFailed on property %d, one or more fields are empty. Check if website format hasn't changed!\nProperty instance: %+v\n", j, p)
			}
			// This is to assure that the limits have the default value if they are not mentioned in the documentation.
			if p.Values.Min == propertyDefaultLimitValue && p.Values.Max != p.Values.Min {
				t.Errorf("\tFailed on property %d, invalid Min and Max.\np.Values.Min = %d, p.Values.Max = %d\n", j, p.Values.Min, p.Values.Max)
			}
			// If the feature isn't going to be added in an upcoming version, but the UpcomingVersion string isn't empty,
			// or vice-versa, it obviously means that gathering the information failed.
			if (p.Upcoming && p.UpcomingVersion == "") || (!p.Upcoming && p.UpcomingVersion != "") {
				t.Errorf("\tFailed on property %d, Upcoming information mismatch.\np.Upcoming = %t, p.UpcomingVersion = %s", j, p.Upcoming, p.UpcomingVersion)
			}
		}
		data, err := json.MarshalIndent(prop, "", "  ")
		if err != nil {
			t.Errorf("Test %d failed, error: %v\n", i, err)
		}
		t.Logf("Test %d data:\n%s\n", i, string(data))
	}
}


v2/filters.go
package v2

import (
	"sort"
	"strings"
)

const (
	filterName     = "name"
	filterType     = "type"
	filterUpcoming = "upcoming"
)

type filters struct {
	contains  map[string]bool
	exactName string
	sortMap   map[string]bool
	typesMap  map[string]bool
	upcoming  string
}

var minecraftTypenames = map[string]bool{
	minecraftStringTypename:  false,
	minecraftIntegerTypename: false,
	minecraftBooleanTypename: false,
}

// validateTypes returns true if filter types are valid, else returns false. If the typemap is valid, the function
// adds missing types, so the filtering works correctly.
func validateTypes(f *filters) bool {
	if f.typesMap == nil {
		return true
	}
	keys := make([]string, 0)
	for k := range f.typesMap {
		keys = append(keys, k)
	}
	for _, k := range keys {
		if _, ok := minecraftTypenames[k]; !ok {
			return false
		}
	}
	prev := f.typesMap[keys[0]]
	for i := 1; i < len(keys); i++ {
		if f.typesMap[keys[i]] != prev {
			return false
		}
	}
	if _, ok := f.typesMap[minecraftBooleanTypename]; !ok {
		f.typesMap[minecraftBooleanTypename] = !prev
	}
	if _, ok := f.typesMap[minecraftIntegerTypename]; !ok {
		f.typesMap[minecraftIntegerTypename] = !prev
	}
	if _, ok := f.typesMap[minecraftStringTypename]; !ok {
		f.typesMap[minecraftStringTypename] = !prev
	}
	return true
}

func (f *filters) filter(p []Property) []Property {
	ret := make([]Property, 0)
	if f.exactName != "" {
		for i := range p {
			if p[i].Name == f.exactName {
				ret = append(ret, p[i])
				break
			}
		}
		return ret
	}
	if f.typesMap == nil && f.contains == nil && f.upcoming == "" {
		return p
	}
	for i := range p {
		if f.typesMap != nil {
			if accept := f.typesMap[p[i].Type]; !accept {
				continue
			}
		}
		if f.contains != nil {
			doesntMatchFilter := false
			for substr, shouldHave := range f.contains {
				if strings.Contains(p[i].Name, substr) == !shouldHave {
					doesntMatchFilter = true
					break
				}
			}
			if doesntMatchFilter {
				continue
			}
		}
		if f.upcoming != "" {
			if (p[i].Upcoming && f.upcoming == "false") || (!p[i].Upcoming && f.upcoming == "true") {
				continue
			}
		}
		ret = append(ret, p[i])
	}
	return ret
}

func (f *filters) sort(p []Property) {
	if order, ok := f.sortMap[filterName]; ok {
		if order {
			sort.Slice(p, func(i, j int) bool {
				return p[i].Name < p[j].Name
			})
		} else {
			sort.Slice(p, func(i, j int) bool {
				return p[i].Name > p[j].Name
			})
		}
	}
	if order, ok := f.sortMap[filterType]; ok {
		if order {
			sort.Slice(p, func(i, j int) bool {
				return p[i].Type < p[j].Type
			})
		} else {
			sort.Slice(p, func(i, j int) bool {
				return p[i].Type > p[j].Type
			})
		}
	}
	if order, ok := f.sortMap[filterUpcoming]; ok {
		if order {
			sort.Slice(p, func(i, j int) bool {
				return p[i].Upcoming == false && p[j].Upcoming == true
			})
		} else {
			sort.Slice(p, func(i, j int) bool {
				return p[i].Upcoming == true && p[j].Upcoming == false
			})
		}
	}
}


v2/scraper.go
package v2

import (
	"io/ioutil"
	"math/big"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"

	"github.com/gocolly/colly"
)

const (
	mathAPI                   = "http://api.mathjs.org/v4/"
	minecraftBooleanTypename  = "boolean"
	minecraftIntegerTypename  = "integer"
	minecraftStringTypename   = "string"
	propertyDefaultLimitValue = -int(^uint32(0)>>1) - 1
	querySelector             = `[data-description="Server properties"]:not([data-description="Server properties"] ~ [data-description="Server properties"]) tr:not(:first-child)`
	regexpLimits              = `(\d+[-–](?:\d+|\(\d+([+\-*/^]\d+[ ]?[+\-*/^][ ]?\d+)?\)))`
	regexpLimitsNotCalculated = `\d+[+\-*/^]\d+ ?[+\-*/^]? ?\d+?`
	wikiLink                  = "https://minecraft.gamepedia.com/Server.properties"
)

var metadata = map[string]interface{}{
	"minecraftBooleanTypename":  minecraftBooleanTypename,
	"minecraftIntegerTypename":  minecraftIntegerTypename,
	"minecraftStringTypename":   minecraftStringTypename,
	"propertyDefaultLimitValue": propertyDefaultLimitValue,
}

// The PropertyValues struct holds all the documented values of a property.
type PropertyValues struct {
	Default  string   `json:"default"`
	Min      int      `json:"min"`
	Max      int      `json:"max"`
	Possible []string `json:"possible"`
}

// The Property struct represents a property with all the information from the official documentation.
type Property struct {
	Name            string         `json:"name"`
	Type            string         `json:"type"`
	Values          PropertyValues `json:"values"`
	Description     string         `json:"description"`
	Upcoming        bool           `json:"upcoming"`
	UpcomingVersion string         `json:"upcomingVersion"`
}

// evaluateMath computes the result of a string expression, using the
// mathjs API: https://api.mathjs.org/
//
// On any error, the return value is always 0.
func evaluateMath(e string) (int, error) {
	res, err := http.Get(mathAPI + "?expr=" + url.QueryEscape(e))
	if err != nil {
		return 0, err
	}
	body, _ := ioutil.ReadAll(res.Body)
	_ = res.Body.Close()
	// Float precision is set to 64, as the input numbers are 64-bit integers.
	calc, _, err := big.ParseFloat(string(body), 10, 64, big.ToNearestEven)
	if err != nil {
		return 0, nil
	}
	calcConv, _ := calc.Int64()
	return int(calcConv), nil
}

// ServerProperties is the main function that scrapes the official documentation based on the passed filters.
func ServerProperties(f *filters) ([]Property, error) {
	var (
		ret      = make([]Property, 0, 50)
		limitErr = error(nil)
	)

	c := colly.NewCollector()
	c.IgnoreRobotsTxt = false

	c.OnHTML(querySelector, func(row *colly.HTMLElement) {
		p := Property{
			Values: PropertyValues{
				Min:      propertyDefaultLimitValue,
				Max:      propertyDefaultLimitValue,
				Possible: []string{},
			},
		}
		row.ForEach(`td`, func(i int, col *colly.HTMLElement) {
			switch i {
			case 0:
				p.Name = col.ChildText(`b`)
				if strings.Contains(col.ChildText(`sup > i > span`), "upcoming") {
					p.Upcoming = true
					p.UpcomingVersion = col.ChildText(`sup > i > a`)
				}
			case 1:
				rawType := strings.TrimSpace(col.Text)
				if rawType == minecraftBooleanTypename {
					p.Type = rawType
					p.Values.Min = 0
					p.Values.Max = 1
				} else if strings.Contains(rawType, minecraftIntegerTypename) {
					p.Type = minecraftIntegerTypename

					// The three regexps are used to get the limits string, split into two separate values and extract
					// the upper limit into an evaluable expression, when it isn't a number.
					// The regexps are always correct. They must be verified to work on all values before modifying them.
					// NOTE: The second, longer dash is included because of an inconsistency on the website.
					r, rs, ru := regexp.MustCompile(regexpLimits), regexp.MustCompile(`([-–])`), regexp.MustCompile(regexpLimitsNotCalculated)

					// l is the limits string, which is found in the rawType.
					l := r.FindString(rawType)
					if l == "" {
						// nothing to do here when the limits string is empty
						return
					}

					// Get the two values.
					values := rs.Split(l, 2)
					// Evaluate the upper limit expression, if necessary
					if e := ru.FindString(values[1]); e != "" {
						result, err := evaluateMath(e)
						if err != nil {
							// This point can be reached only if the mathjs API is down.
							// Exiting the function assures that Min and Max are both propertyDefaultLimitValue or
							// both assigned another value
							limitErr = err
							return
						}
						p.Values.Max = result
					} else {
						p.Values.Max, _ = strconv.Atoi(values[1])
					}
					p.Values.Min, _ = strconv.Atoi(values[0])
				} else if rawType == minecraftStringTypename {
					p.Type = minecraftStringTypename
				}
			case 2:
				p.Values.Default = strings.TrimSpace(col.Text)
			case 3:
				p.Description = strings.TrimSpace(col.Text)
				// Loop through each significant value, appending them to the Possible
				// field of the Values field of the p instance. If property is boolean,
				// this is unnecessary.
				if p.Type == minecraftBooleanTypename {
					return
				}
				col.ForEach(`dl dd`, func(_ int, val *colly.HTMLElement) {
					t := val.ChildText(`b:first-child`)
					// Don't append empty values, as they are coded in the markdown of the
					// website, and also don't append Note notices, as they are inside the same
					// tag structure in the HTML
					if t == "" || t == "Note:" {
						return
					}
					p.Values.Possible = append(p.Values.Possible, t)
				})
			}
		})
		ret = append(ret, p)
	})
	err := c.Visit(wikiLink)
	if err != nil {
		return nil, err
	}
	if limitErr != nil {
		return nil, limitErr
	}

	if f != nil {
		ret = f.filter(ret)
		f.sort(ret)
	}

	return ret, nil
}


runapi/main.go
package main

import (
	"log"
	"net/http"
	"time"

	"github.com/NYTimes/gziphandler"

	cache "github.com/victorspringer/http-cache"

	"github.com/victorspringer/http-cache/adapter/memory"

	"github.com/gorilla/mux"
	spa "github.com/tmaxmax/serverpropertiesapi"
)

func main() {
	memcached, err := memory.NewAdapter(
		memory.AdapterWithAlgorithm(memory.LRU),
		memory.AdapterWithCapacity(10000000),
	)
	if err != nil {
		log.Fatalln(err)
	}
	cacheClient, err := cache.NewClient(
		cache.ClientWithAdapter(memcached),
		cache.ClientWithTTL(24*time.Hour),
		cache.ClientWithRefreshKey("opn"),
	)
	if err != nil {
		log.Fatalln(err)
	}

	// main router
	r := mux.NewRouter()
	// server.properties API sub router
	sprop := r.PathPrefix("/v1/serverproperties").Subrouter()

	sprop.Handle("", cacheClient.Middleware(gziphandler.GzipHandler(http.HandlerFunc(spa.GetAllProperties)))).Methods(http.MethodGet)
	sprop.Handle("/{key}", cacheClient.Middleware(gziphandler.GzipHandler(http.HandlerFunc(spa.GetProperty)))).Methods(http.MethodGet)
	sprop.Handle("/meta/", cacheClient.Middleware(gziphandler.GzipHandler(http.HandlerFunc(spa.GetMetadata)))).Methods(http.MethodGet)

	sprop.Handle("", gziphandler.GzipHandler(http.HandlerFunc(spa.MethodNotAllowedHandler(http.MethodGet))))
	sprop.Handle("/{key}", gziphandler.GzipHandler(http.HandlerFunc(spa.MethodNotAllowedHandler(http.MethodGet))))
	sprop.Handle("/meta/", gziphandler.GzipHandler(http.HandlerFunc(spa.MethodNotAllowedHandler(http.MethodGet))))

	log.Fatalln(http.ListenAndServe(":80", r))
}


v2/serverpropertiesapi.go
package v2

import (
	"encoding/json"
	"net/http"

	"github.com/gorilla/mux"
)

var (
	apiErrorInvalidTypeQuery = rfc7807{
		errType: "",
		Title:   "Invalid type query",
		Status:  http.StatusBadRequest,
		Detail:  "The type query filter contains mixed allowed and disallowed values",
	}
)

func GetServerProperties() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if writeError(w, checkRequest(r)) {
			return
		}

		filtersQuery := r.URL.Query()
		f := &filters{
			contains:  obtainMap(makeCompleteArray(filtersQuery["contains"]), '!'),
			exactName: "",
			sortMap:   obtainMap(makeCompleteArray(filtersQuery["sort"]), '-'),
			typesMap:  obtainMap(makeCompleteArray(filtersQuery["types"]), '!'),
			upcoming:  filtersQuery.Get("upcoming"),
		}
		if !validateTypes(f) {
			writeError(w, apiErrorInvalidTypeQuery)
			return
		}

		prop, err := ServerProperties(f)
		if err != nil {
			writeError(w, apiErrorServerError)
			return
		}

		data, _ := json.MarshalIndent(prop, "", "  ")
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		w.Write(data)
	})
}

func GetServerProperty() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if writeError(w, checkRequest(r)) {
			return
		}

		p, err := ServerProperties(&filters{
			exactName: mux.Vars(r)["key"],
		})
		if err != nil {
			writeError(w, apiErrorServerError)
			return
		}
		if len(p) == 0 {
			writeError(w, apiErrorPropertyNotFound)
			return
		}
		data, _ := json.MarshalIndent(p[0], "", "  ")
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		w.Write(data)
	})
}

func GetMetadata() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if writeError(w, checkRequest(r)) {
			return
		}

		data, err := json.MarshalIndent(metadata, "", "  ")
		if err != nil {
			writeError(w, apiErrorServerError)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		w.Write(data)
	})
}


api.go
package serverpropertiesapi

import (
	"encoding/json"
	"net/http"
	"regexp"
	"strings"

	"github.com/gorilla/mux"
)

type Error struct {
	httpCode int
	Error    string `json:"error"`
	Retry    bool   `json:"retry"`
}

var (
	internalServerError = Error{
		httpCode: http.StatusInternalServerError,
		Error:    "500 Internal Server Error, failed to get information",
		Retry:    false,
	}
	metaInformation = map[string]interface{}{
		"minecraftBooleanTypename":  minecraftBooleanTypename,
		"minecraftIntegerTypename":  minecraftIntegerTypename,
		"minecraftStringTypename":   minecraftStringType,
		"propertyDefaultLimitValue": propertyDefaultLimitValue,
	}
)

// makeCompleteArray transforms a slice that contains elements which are made of comma separated strings
// into a slice that contains all the elements on a different position
//
// Example: ["a", "b,c", "d,e,f"] -> ["a", "b", "c", "d", "e", "f"]
func makeCompleteArray(a []string) []string {
	b := make([]string, len(a))
	copy(b, a)
	for i := 0; i < len(b); i++ {
		c := strings.Split(b[i], ",")
		if len(c) > 1 {
			b = append(b[:i], append(c, b[i+1:]...)...)
			i += len(c) - 1
		}
	}
	return b
}

// checkRequest checks the request header if the data necessary to make an API request is available
// It returns an Error slice, which contains all the problems found in the request.
func checkRequest(r *http.Request) []Error {
	var ret []Error

	// Get the accepted formats string
	acceptHeader := r.Header.Get("Accept")

	// Construct the split regex, used to split the header Accept string value
	// into all the requested formats
	split := regexp.MustCompile(`,[ ]?|;[ ]?[qv]=([^,;]+)`)

	// Check if the client requests data in a supported format. If no format is provided,
	// or any format is accepted, application/json is implied.
	accept, i := split.Split(acceptHeader, -1), 0
	for ; i < len(accept); i++ {
		if accept[i] == "*/*" || accept[i] == "application/json" {
			break
		}
	}

	if len(accept) != 0 && i == len(accept) {
		ret = append(ret, Error{
			httpCode: http.StatusNotImplemented,
			Error:    "501 Not Implemented, API does not support " + acceptHeader + " format",
			Retry:    false,
		})
	}

	return ret
}

// writeErrors writes all the errors found until the point this function is called to the response.
// The response status code will be the httpCode of the first error.
//
// Always check if there are any errors before calling writeErrors!
func writeErrors(w http.ResponseWriter, errors ...Error) {
	data, _ := json.MarshalIndent(struct {
		Errors []Error `json:"errors"`
	}{errors}, "", "  ")
	w.WriteHeader(errors[0].httpCode)
	w.Write(data)
}

// GetAllProperties is a higher order function that has as parameter the a Property slice,
// which is used to send the information the client requests. It returns a function that
// shall be used as a handler for GET requests of all properties.
func GetAllProperties(w http.ResponseWriter, r *http.Request) {
	w.Header().Add("Content-Type", "application/json")

	errors := checkRequest(r)
	if len(errors) != 0 {
		writeErrors(w, errors...)
		return
	}

	filters := r.URL.Query()
	opt := Options{
		Contains:  makeCompleteArray(filters["contains"]),
		exactName: "",
		Types:     makeCompleteArray(filters["types"]),
		Upcoming:  filters.Get("upcoming"),
	}
	if !opt.Valid() {
		writeErrors(w, Error{
			httpCode: http.StatusBadRequest,
			Error:    "400 Bad Request, options are not valid",
			Retry:    false,
		})
		return
	}
	p, err := ServerProperties(opt)
	if err != nil {
		writeErrors(w, internalServerError)
		return
	}

	data, _ := json.MarshalIndent(struct {
		Options    Options    `json:"options"`
		Properties []Property `json:"properties"`
	}{Options: opt, Properties: p}, "", "  ")
	w.WriteHeader(http.StatusOK)
	w.Write(data)
}

func GetProperty(w http.ResponseWriter, r *http.Request) {
	w.Header().Add("Content-Type", "application/json")

	errors := checkRequest(r)
	if len(errors) != 0 {
		writeErrors(w, errors...)
		return
	}

	// Get the requested key
	pathParams := mux.Vars(r)
	name := pathParams["key"]

	// Find the Property instance with the given name
	p, err := ServerProperty(name)
	if err != nil {
		if err.Error() == "not found" {
			writeErrors(w, Error{
				httpCode: http.StatusNotFound,
				Error:    "404 Not Found, key \"" + name + "\" doesn't exist",
				Retry:    false,
			})
			return
		}
		writeErrors(w, internalServerError)
		return
	}

	// If this point is reached, there are no errors.
	// Send the Property instance to the client.
	data, _ := json.MarshalIndent(p, "", "  ")
	w.WriteHeader(http.StatusOK)
	w.Write(data)
}

func GetMetadata(w http.ResponseWriter, r *http.Request) {
	w.Header().Add("Content-Type", "application/json")

	errors := checkRequest(r)
	if len(errors) != 0 {
		writeErrors(w, errors...)
		return
	}

	data, _ := json.MarshalIndent(struct {
		Meta map[string]interface{} `json:"meta"`
	}{metaInformation}, "", "  ")
	w.WriteHeader(http.StatusOK)
	w.Write(data)
}

func MethodNotAllowedHandler(allowed ...string) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		allowedMethods := strings.Join(allowed, ", ")
		errors := checkRequest(r)
		errors = append(errors, Error{
			httpCode: http.StatusMethodNotAllowed,
			Error:    "405 Status Method Not Allowed, can't use " + r.Method,
			Retry:    false,
		})
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Allowed", allowedMethods)
		writeErrors(w, errors...)
	}
}
